File: S20165190/HW13/hw13_1/hw13_1.cpp
/*
 * hw12.cpp
 *
 *  Created on: 2018. 11. 25.
 *      Author: han
 */

#include "LA.h"
#include "const.cpp"

double **SCeq(double Vapp);
double Ber(double x);

int main(int argc, char* argv[]){
	double **result, *phi, *elec, *J, Vapp=0.0;
	N = atoi(argv[2]);
	J = new double[N-1];
	if(*argv[1] == 's' || *argv[1] == 'S') {
		Hdop = 5e25;
		Ldop = 2e23;
		Length = 120e-9;
		interface1 = (N-1)/3;
		interface2 = (N-1)*2/3;
	}
	else{
		interface1 = (N-1)/6;
		interface2 = (N-1)*5/6;
	}
	termx = Length/(double)(N-1);
	coef = termx*termx*q/eps0;

	for(int j = 0; j < 11; j++)
	{
		result = SCeq(Vapp);
		phi = *result;
		elec = *(result+1);
	//	for(int i = 0; i < N; i++) cout << elec[i] << '\t' << phi[i] << endl;
		for(int i = 0; i < N-1; i++) J[i] = q*Dn/termx*(elec[i+1]*Ber((phi[i+1]-phi[i])/thermal)-elec[i]*Ber((phi[i]-phi[i+1])/thermal));

		cout << Vapp << " ";
		for(int i = 0; i < N-1; i++) cout << J[i] << " ";
		cout << endl;
		Vapp += 0.05;
		delete(phi, elec, result);
	}
	return 0;
}


double **SCeq(double Vapp)
{

	LA jaco(2*N);
	double **result = new double*[2];

	double *res = new double[2*N], *update, *elec = new double[N], *phi = new double[N];
	double *Ndon = new double[N], *Cvector = new double[2*N], *Rvector = new double[2*N];
	double n_av, dphidx, delecdx, Jn, sum;
	result[0] = phi; result[1] = elec;


	for(int i = 0; i < N; i++)
	{
		if(i <= interface1 || i >= interface2) Ndon[i] = Hdop;
		else Ndon[i] = Ldop;

		phi[i] = thermal*log(Ndon[i]/ni);
		elec[i] = ni*exp(phi[i]/thermal);
	}

	for(int r = 0; r < VN; r++)
	{
		for(int i = 0; i < 2*N; i++) {
			res[i] = 0;
			for(int j = 0; j < 2*N; j++) jaco(i,j) = 0;
		}

		res[0] = phi[0] - thermal*log(Ndon[0]/ni);
		jaco(0,0) = 1.0;

		for(int i = 1; i < N-1; i++)
		{
			res[2*i] = eSi*(phi[i+1] - 2*phi[i] + phi[i-1]) + coef*(Ndon[i]-elec[i]);
			jaco(2*i,2*i-2) = eSi;
			jaco(2*i,2*i) = -2.0*eSi;
			jaco(2*i,2*i+2) = eSi;
			jaco(2*i,2*i+1) = -coef;
		}
		res[2*N-2] = phi[N-1] - thermal*log(Ndon[N-1]/ni) - Vapp;
		jaco(2*N-2, 2*N-2) = 1.0;

		res[1] = elec[0] - Ndon[0];
		jaco(1,1) = 1.0;
		for(int i = 1; i < N-1; i++)
		{
			n_av = 0.5*(elec[i+1]+elec[i]);
			dphidx = (phi[i+1]-phi[i])/termx;
			delecdx = (elec[i+1]-elec[i])/termx;
			Jn = n_av * dphidx - thermal * delecdx;
			res[2*i+1] += Jn;
			jaco(2*i+1, 2*i+3) += 0.5*dphidx - thermal / termx;
			jaco(2*i+1, 2*i+1) += 0.5*dphidx + thermal / termx;
			jaco(2*i+1, 2*i+2) += n_av / termx;
			jaco(2*i+1, 2*i) += - n_av / termx;
			res[2*i+3] += -Jn;
			jaco(2*i+3, 2*i+3) += -0.5*dphidx + thermal / termx;
			jaco(2*i+3, 2*i+1) += -0.5*dphidx - thermal / termx;
			jaco(2*i+3, 2*i+2) += - n_av / termx;
			jaco(2*i+3, 2*i) += n_av / termx;
		}
		res[2*N-1] = elec[N-1] - Ndon[N-1];
		jaco(2*N-1,2*N-1) = 1.0;

		for(int i = 0; i < N; i++)
		{
			Cvector[2*i] = thermal;
			Cvector[2*i+1] = sqrt((double)2*interface1*(Hdop*Hdop)+(double)(N-interface1-interface2)*(Ldop*Ldop));
		}
		jaco.Rdiag(Cvector);
		sum = jaco.abssum(1);
		for(int i = 0; i < 2*N; i++) Rvector[i] = 1.0/sum;
		jaco.Ldiag(Rvector);
		for(int i = 0; i < 2*N; i++) res[i] *= Rvector[i];
		update = -jaco / (res);
		for(int i = 0; i < 2*N; i++) update[i] *= Cvector[i];

		for(int i = 0; i < N; i++) phi[i] += update[2*i];
		for(int i = 0; i < N; i++) elec[i] += update[2*i+1];
	}

	jaco.~LA();
	delete(res, Ndon, Cvector, Rvector);
	return result;
}

double Ber(double x){ return x/(exp(x)-1.0); }

File: s20174003/HomeWork13/hw13_1.m
close all;
clear all;

Structure_sel = 1; % 0 is long, 1 is short structure.
q = 1.602192e-19;
k_B = 1.380662e-23;
eps0 = 8.854187817e-12; % Vaccum permittivity (F/m)
T = 300.0;
thermal = k_B*T/q;
Dn = 0.1;

%Deltax = 1e-9; % 1nm spacing


if(Structure_sel==0)%% Long structure
    N = 601;
    Deltax = 1e-9;
    x_12 = 101; %At x=100nm
    x_23 = 501; %At x=500nm
else%%Short structure
    N = 121;
    Deltax =1e-9;
    xs_12 = 41; %At x=40nm
    xs_23 = 81; %At x=80nm
end

x = Deltax*transpose([0:N-1]); %real space (m)
eps_si = 11.7;
eps_ox = 3.9;

if(Structure_sel==0)
    Ndon = 2e21*ones(N,1); %2e15 /cm^3;
    Ndon(1:x_12,1) = 5e23; %5e17 /cm^3;
    Ndon(x_23:N,1) = 5e23; %5e17 /cm^3;
else
    Ndon = 2e23*ones(N,1); %2e17 /cm^3;
    Ndon(1:xs_12,1) = 5e25; %5e19 /cm^3;
    Ndon(xs_23:N,1) = 5e25; %5e19 /cm^3;
end
ni = 1.075e16; %1.075e10 /cm^3
coef = Deltax*Deltax*q/eps0;

phi = zeros(N,1);
phi(:,1) = thermal*log(Ndon(:,1)/ni);
for newton=1:10
    res = zeros(N,1);
    Jaco = sparse(N,N);
    res(1,1) = phi(1,1) - thermal*log(Ndon(1,1)/ni);
    Jaco(1,1) = 1.0;
    for iter=2:N-1
        res(iter,1) = eps_si*(phi(iter+1,1) - 2*phi(iter,1) + phi(iter-1,1));
        Jaco(iter,iter-1) = eps_si;
        Jaco(iter,iter) = -2*eps_si;
        Jaco(iter,iter+1) = eps_si;
    end
    res(N,1) = phi(N,1) - thermal*log(Ndon(N,1)/ni);
    Jaco(N,N) = 1.0;
    for iter=2:N-1
        res(iter,1) = res(iter,1) - coef*(-Ndon(iter,1) + ni*exp(phi(iter,1)/thermal));
        Jaco(iter,iter) = Jaco(iter,iter) - coef*ni*exp(phi(iter,1)/thermal)/thermal;
    end
    update = Jaco \ (-res);
    phi = phi + update;
    %norm(update,inf)
end



elec = zeros(N,1);
elec = ni*exp(phi/thermal);
J = zeros(N-1,1);
sub_iter = 0;
phi = zeros(N,1);
phi(:,1) = thermal*log(Ndon(:,1)/ni);
for bias=0:10
    V_applied = 0.05*bias;
    sub_iter = sub_iter +1;
    for newton=1:10
        res = zeros(2*N,1);
        Jaco = sparse(2*N,2*N);
        res(1,1) = phi(1,1) - thermal*log(Ndon(1,1)/ni);
        Jaco(1,1) = 1.0;
        for iter=2:N-1
            res(2*iter-1,1) = eps_si*(phi(iter+1,1) - 2*phi(iter,1) + phi(iter-1,1)) + coef*(Ndon(iter,1)-elec(iter,1));
            Jaco(2*iter-1,2*iter+1) = eps_si;
            Jaco(2*iter-1,2*iter-1) = -2*eps_si;
            Jaco(2*iter-1,2*iter-3) = eps_si;
            Jaco(2*iter-1,2*iter) = -coef;
        end
        res(2*N-1,1) = phi(N,1) - thermal*log(Ndon(N,1)/ni) - V_applied;
        Jaco(2*N-1,2*N-1) = 1.0;

        for iter=1:N-1
            n_av = 0.5*(elec(iter+1,1)+elec(iter,1));
            dphidx = (phi(iter+1,1) - phi(iter,1))/Deltax;
            delecdx = (elec(iter+1,1) - elec(iter,1))/Deltax;
            Jn = n_av * dphidx - thermal * delecdx;
            res(2*iter,1) = res(2*iter,1) + Jn;
            Jaco(2*iter,2*iter+2) = Jaco(2*iter,2*iter+2) + 0.5*dphidx - thermal / Deltax;
            Jaco(2*iter,2*iter) = Jaco(2*iter,2*iter) + 0.5*dphidx + thermal / Deltax;
            Jaco(2*iter,2*iter+1) = Jaco(2*iter,2*iter+1) + n_av / Deltax;
            Jaco(2*iter,2*iter-1) = Jaco(2*iter,2*iter-1) - n_av / Deltax;
            res(2*iter+2,1) = res(2*iter+2,1) - Jn;
            Jaco(2*iter+2,2*iter+2) = Jaco(2*iter+2,2*iter+2) - 0.5*dphidx + thermal /Deltax;
            Jaco(2*iter+2,2*iter) = Jaco(2*iter+2,2*iter) - 0.5*dphidx - thermal / Deltax;
            Jaco(2*iter+2,2*iter+1) = Jaco(2*iter+2,2*iter+1) - n_av / Deltax;
            Jaco(2*iter+2,2*iter-1) = Jaco(2*iter+2,2*iter-1) + n_av / Deltax;
        end

        res(2,1) = elec(1,1) - Ndon(1,1);
        Jaco(2,:) = 0;
        Jaco(2,2) = 1.0;
        res(2*N,1) = elec(N,1) - Ndon(N,1);
        Jaco(2*N,:) = 0;
        Jaco(2*N,2*N) = 1.0;



        Cvector = zeros(2*N,1);
        Cvector(1:2:2*N-1,1) = thermal;
        Cvector(2:2:2*N,1) = max(abs(Ndon));
        Cmatrix = spdiags(Cvector,0,2*N,2*N);
        Jaco_scaled = Jaco * Cmatrix;
        Rvector = 1./sum(abs(Jaco_scaled),2);
        Rmatrix = spdiags(Rvector,0,2*N,2*N);
        Jaco_scaled = Rmatrix * Jaco_scaled;
        res_scaled = Rmatrix * res;
        update_scaled = Jaco_scaled \ (-res_scaled);
        update =Cmatrix * update_scaled;



        %update = Jaco \ (-res);
        phi = phi + update(1:2:2*N-1,1);
        elec = elec + update(2:2:2*N,1);
        %norm(update(1:2:2*N-1,1),inf)
    end


    for iteriter=1:N-1
        Bernoulli_x = (phi(iteriter+1,1)-phi(iteriter,1))/thermal;
        J(iteriter,1) = ((-q*Dn)/(Deltax))*(elec(iteriter+1,1)*(Bernoulli_x/(exp(Bernoulli_x)-1))-elec(iteriter,1)*(-Bernoulli_x/(exp(-Bernoulli_x)-1))); 
    end
    Current(sub_iter,1) = J(end,1);
end

V_sweep = 0:0.05:0.5;


figure(1)

semilogy(V_sweep,Current);
xlabel('Drain Voltage');
ylabel('Current Density at Drain[A/m^2]')
grid on;
